ui/imports/utils/SubscriptionBroker.qml

```diff
@@ -0,0 +1,196 @@
+import QtQuick 2.15
+
+import utils 1.0
+
+//This is a helper component that is used to batch requests and send them periodically
+//It is used to reduce the number of requests sent to the server and notify different subscribers of the same request
+//It is used by the Subscription component
+QtObject {
+    id: root
+
+    signal request(string topic)
+    function response(topic, responseObj) {
+        d.onResponse(topic, responseObj)
+    }
+    function subscribe(subscription) {
+        d.subscribe(subscription)
+    }
+    function unsubscribe(subscription) {
+        d.unsubscribe(subscription)
+    }
+
+
+    readonly property QtObject d: QtObject {
+        //Mapping subscriptionId to subscription object
+        //subscriptionId is a string and represents the id of the subscription
+        //E.g. "subscriptionId": {subscription: subscriptionObject, topic: "topic"}
+        //The purpose of this mapping is to keep track of the subscriptions and their topics
+        readonly property var managedSubscriptions: ({})
+
+        //Mapping topic to subscriptionIds and request data
+        //topic is a string and represents the topic of the subscription
+        //E.g. "topic": {requestTimestamp: 0, requestInterval: 1000, subscriptions: ["subscriptionId1", "subscriptionId2"], response: null}
+        readonly property var topics: ({})
+        property int topicsCount: 0 //helper property to track change events
+
+        readonly property int requestInterval: {
+            //dependency:
+            d.topicsCount
+            const topicInfos = Object.values(d.topics)
+            if(!topicsCount || topicInfos.length === 0) return 0
+            
+            return topicInfos.reduce(function(minInterval, topicInfo) { return Math.min(minInterval, topicInfo.requestInterval) }, Number.MAX_SAFE_INTEGER)
```

Comment:
I'm wondering what's the behavior when we have e.g. 2 subscriptions:
1. First subscription has requestInterval = 10s
2. Second subscription has requestInterval = 9s

The first request is sent for both at start.

The second request:
 -> For the second subscription the second request will be sent after 9s
 -> For the first subscription the second request will be sent after 18s

Do I understand it correctly? If it works this way, the broker may be unreliable when using various intervals.

Reply:
Wdyt about creating some simple example for the broker in a storybook page? As we don't have unit tests, it could present some internal details and allow to quickly understand the idea.

Reply:
> I'm wondering what's the behavior when we have e.g. 2 subscriptions:
> 
> 1. First subscription has requestInterval = 10s
> 2. Second subscription has requestInterval = 9s
> 
> The first request is sent for both at start.
> 
> The second request: -> For the second subscription the second request will be sent after 9s -> For the first subscription the second request will be sent after 18s
> 
> Do I understand it correctly? If it works this way, the broker may be unreliable when using various intervals.

Fixed! Thanks for pointing that out.


> Wdyt about creating some simple example for the broker in a storybook page? As we don't have unit tests, it could present some internal details and allow to quickly understand the idea.

It's a good point! Now that we have the unit tests in the CI I've added qml tests for the SubscriptionBroker.qml. Was a good occasion to spot corner cases 💪